packages/slate/src/core/normalize-node.ts

```diff
@@ -1,123 +1,134 @@
-  // Since we'll be applying operations while iterating, keep track of an
-  // index that accounts for any added/removed nodes.
-  let n = 0
+  // Determine whether the node should have only block or only inline children.
+  // - The editor should have only block children.
+  // - Inline elements should have only inline children.
+  // - Elements that begin with a text child or an inline element child should have only inline children.
+  // - All other elements should have only block children.
+  const shouldHaveInlines =
+    !(element === editor) &&
+    (editor.isInline(element) ||
+      Text.isText(element.children[0]) ||
+      editor.isInline(element.children[0]))
 
-  for (let i = 0; i < node.children.length; i++, n++) {
-    const currentNode = Node.get(editor, path)
-    if (Text.isText(currentNode)) continue
-    const child = currentNode.children[n] as Descendant
-    const prev = currentNode.children[n - 1] as Descendant
-    const isLast = i === node.children.length - 1
-    const isInlineOrText =
-      Text.isText(child) || (Element.isElement(child) && editor.isInline(child))
+  if (shouldHaveInlines) {
+    // Since we'll be applying operations while iterating, we also modify `n` when adding/removing nodes.
+    for (let n = 0; n < element.children.length; n++) {
+      const child = element.children[n]
+      const prev = element.children[n - 1] as Descendant | undefined
 
-    // Only allow block nodes in the top-level children and parent blocks
-    // that only contain block nodes. Similarly, only allow inline nodes in
-    // other inline nodes, or parent blocks that only contain inlines and
-    // text.
-    if (isInlineOrText !== shouldHaveInlines) {
-      if (isInlineOrText) {
-        if (options?.fallbackElement) {
-          Transforms.wrapNodes(editor, options.fallbackElement(), {
-            at: path.concat(n),
-            voids: true,
-          })
-        } else {
-          Transforms.removeNodes(editor, { at: path.concat(n), voids: true })
+      if (Text.isText(child)) {
+        if (prev != null && Text.isText(prev)) {
+          // Merge adjacent text nodes that are empty or match.
+          if (child.text === '') {
+            Transforms.removeNodes(editor, {
+              at: path.concat(n),
+              voids: true,
+            })
+            element = Node.get(editor, path) as Element
+            n--
+          } else if (prev.text === '') {
+            Transforms.removeNodes(editor, {
+              at: path.concat(n - 1),
+              voids: true,
+            })
+            element = Node.get(editor, path) as Element
+            n--
+          } else if (Text.equals(child, prev, { loose: true })) {
+            Transforms.mergeNodes(editor, { at: path.concat(n), voids: true })
+            element = Node.get(editor, path) as Element
+            n--
+          }
         }
       } else {
-        Transforms.unwrapNodes(editor, { at: path.concat(n), voids: true })
-      }
-      n--
```

Comment:
pretty sure there's an invisible bug if this runs. we add some number of new child nodes and step back to read them--but we still only run the loop a number of times equal to the number of original children (everywhere else we add a child we skip over it and num iterations stay the same)

I dont think it caused issues due to the multi-pass nature that `normalizeNode` gets called but we can avoid running it more times than we need to by accounting for it.